docs/01-app/01-getting-started/15-backend-for-frontend.mdx

```diff
@@ -0,0 +1,1052 @@
+    )
+  }
+}
+```
+
+You can also proxy requests using `middleware`:
+
+```ts filename="middleware.ts" switcher
+import { NextResponse } from 'next/server'
+
+export function middleware(request: Request) {
+  if (request.nextUrl.pathname === '/proxy-this-path') {
+    const rewriteUrl = new URL('https://nextjs.org')
+    return NextResponse.rewrite(rewriteUrl)
+  }
+}
+```
+
+```js filename="middleware.js" switcher
+import { NextResponse } from 'next/server'
+
+export function middleware(request) {
+  if (request.nextUrl.pathname === '/proxy-this-path') {
+    const rewriteUrl = new URL('https://nextjs.org')
+    return NextResponse.rewrite(rewriteUrl)
+  }
+}
+```
+
+Another type of response `middleware` can produce are redirects:
+
+```ts filename="middleware.ts" switcher
+import { NextResponse } from 'next/server'
+
+export function middleware(request: Request) {
+  if (request.nextUrl.pathname === '/v1/docs') {
+    request.nextUrl.pathname = '/v2/docs'
+    return NextResponse.redirect(request.nextUrl)
+  }
+}
+```
+
+```js filename="middleware.js" switcher
+import { NextResponse } from 'next/server'
+
+export function middleware(request) {
+  if (request.nextUrl.pathname === '/v1/docs') {
+    request.nextUrl.pathname = '/v2/docs'
+    return NextResponse.redirect(request.nextUrl)
+  }
+}
+```
+
+## Security
+
+### Rate Limiting
+
+You can implement rate limiting in your Next.js backend. In addition to code-based checks, enable any rate limiting features provided by your host.
+
+```ts filename="/app/resource/route.ts" switcher
+import { NextResponse } from 'next/server'
+import { checkRateLimit } from '@/lib/rate-limit'
+
+export async function POST(request: Request) {
+  const { rateLimited } = await checkRateLimit(request)
+
+  if (rateLimited) {
+    return NextResponse.json({ error: 'Rate limit exceeded' }, { status: 429 })
+  }
+
+  return new Response(null, { status: 204 })
+}
+```
+
+```js filename="/app/resource/route.js" switcher
+import { NextResponse } from 'next/server'
+import { checkRateLimit } from '@/lib/rate-limit'
+
+export async function POST(request) {
+  const { rateLimited } = await checkRateLimit(request)
+
+  if (rateLimited) {
+    return NextResponse.json({ error: 'Rate limit exceeded' }, { status: 429 })
+  }
+
+  return new Response(null, { status: 204 })
+}
+```
+
+### Verify Payloads
+
+Never trust incoming request data. Validate content type and size, and sanitize against XSS before use.
+
+Use timeouts to prevent abuse and protect server resources.
+
+Store user-generated static assets in dedicated services. When possible, upload them from the browser and store the returned URI in your database to reduce request size.
+
+### Access to protected resources
+
+Always verify credentials before granting access. Do not rely on middleware alone for authentication and authorization.
+
+Remove sensitive or unnecessary data from responses and backend logs.
+
+Rotate credentials and API keys regularly.
```

Comment:
Not sure if we want to expand further here, or just link to the Authentication guide.

I have considered a similar example to the Authentication guide, but along the lines of:

```ts
import { OTP } from 'lib/otp';

export const DELETE = async (req) => {
   const isValidReq = validateReq(req);

   if (isValidReq) {
      const otpSuccess = await OTP(req);
      // then allow to DELETE the resource
   }
}

```

The idea being that some operations, might require stronger protection.

I think it might overlap with the security paper.